db/db_impl/db_impl_open.cc

```diff
@@ -529,7 +530,30 @@ Status DBImpl::Recover(
   Status s;
   bool missing_table_file = false;
   if (!immutable_db_options_.best_efforts_recovery) {
-    s = versions_->Recover(column_families, read_only, &db_id_);
+    // Status of reading the descriptor file
+    Status desc_status;
+    s = versions_->Recover(column_families, read_only, &db_id_,
+                           /*no_error_if_files_missing=*/false, is_retry,
+                           &desc_status);
+    desc_status.PermitUncheckedError();
+    if (can_retry) {
+      // If we're opening for the first time and the failure is likely due to
+      // a corrupt MANIFEST file (could result in either the log::Reader
+      // detecting a corrupt record, or SST files not found error due to
+      // discarding badly formed tail records)
+      if (!is_retry && (desc_status.IsCorruption() || s.IsNotFound()) &&
```

Comment:
How about `s.IsCorruption()`?

In particular I am thinking about the case of empty MANIFEST, which fails to open with "Corruption: no log_file_number, next_file_number, last_sequence entry in MANIFEST  The file ./tmp-db/MANIFEST-000005 may be corrupted."

I thought for those cases that we saw, rereading with reconstruction helped, but I actually don't know why. Do you? In particular I'm confused about whether the file size was zero in the metadata because it feels like that would cause us to not find the data whether it's a reconstruction read or not.

Reply:
> I thought for those cases that we saw, rereading with reconstruction helped, but I actually don't know why

Might have misremembered. Would be fine with me to say handling wrong file size metadata is out of scope.

My `s.IsCorruption()` example may still be relevant, for example, for certain corruptions at the start of the file.

Reply:
Makes sense. My initial thought was to narrow it down to MANIFEST file level corruption, but yeah, its possible some may show up as `s.IsCorruption()`. We may end up casting a wider net, for example, catching SST corruption detected by `LoadTables`, but there's no downside to retrying.